### PR TITLE
Add concrete dropout

### PR DESCRIPTION
**Pull request by jimfleming**

### CODE REVIEW

File: pyoneer/layers/dropout_impl.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        self.dropout_regularization = dropout_regularization
+        self.noise_shape = noise_shape
+        self.seed = seed
+        self.supports_masking = True
+        self.rate_logit = None
+
+    def build(self, input_shape):
+        super(ConcreteDropout, self).build(input_shape)
+
+        initial_rate_norm = np.log(self.initial_rate) - np.log(1.0 - self.initial_rate)
```
Comment (wenkesj) [inline pyoneer/layers/dropout_impl.py]: Can this be implemented with tensorflow to support tensors and numpy arrays?

File: pyoneer/layers/dropout_impl.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    def get_config(self):
+        config = {
+            "initial_rate": self.initial_rate,
+            "kernel_regularization": self.kernel_regularization,
+            "dropout_regularization": self.dropout_regularization,
+            "noise_shape": self.noise_shape,
+            "seed": self.seed,
+        }
+        base_config = super(Dropout, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
```
Comment (djsaunde) [inline pyoneer/layers/dropout_impl.py]: Can this be replaced with `return {**base_config, **config}`?

File: pyoneer/layers/dropout_impl.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        self.rate_logit = self.add_weight(
+            name="rate_logit", shape=(), initializer=rate_initializer, trainable=True
+        )
+
+        if self.noise_shape is None:
+            self.noise_shape = input_shape
+
+        input_dim = np.prod(self.noise_shape[1:])
+
+        kernel_regularizer = (
```
Comment (wenkesj) [inline pyoneer/layers/dropout_impl.py]: Can this be implemented as a [tf.keras.reguralizers.Regularizer](https://www.tensorflow.org/api_docs/python/tf/keras/regularizers/Regularizer) and the model should use this (you could add it in the module here)? It might make this semantically easier to understand

File: pyoneer/layers/dropout_impl.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    @property
+    def rate(self):
+        assert (
+            self.rate_logit is not None
+        ), "the layer must be called before accessing rate"
+        return tf.math.sigmoid(self.rate_logit)
+
+    def concrete_dropout(self, inputs):
+        epsilon = tf.keras.backend.cast_to_floatx(tf.keras.backend.epsilon())
+        temperature = 0.1
```
Comment (djsaunde) [inline pyoneer/layers/dropout_impl.py]: Why is `temperature` fixed? Can it be a argument to `__init__` instead? [The Concrete distribution paper](https://openreview.net/forum?id=S1jE5L5gl) indicates temperature can be chosen in (0, inf).

File: pyoneer/layers/dropout_impl.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        return self.layer.compute_output_shape(input_shape)
+
+    @property
+    def rate(self):
+        assert (
+            self.rate_logit is not None
+        ), "the layer must be called before accessing rate"
+        return tf.math.sigmoid(self.rate_logit)
+
+    def concrete_dropout(self, inputs):
```
Comment (wenkesj) [inline pyoneer/layers/dropout_impl.py]: Should this be implemented as a function that takes the `self` attributes as arguments? My reasoning behind this is that the layer itself could just be a thin-layer above some low-level functions like the `concrete_dropout(inputs, rate, noise_shape, seed, epsilon)` for those who don't necessarily use the keras API?

File: pyoneer/layers/dropout_test.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+
+from pyoneer.layers.dropout_impl import ConcreteDropout
+
+
+class DropoutTest(tf.test.TestCase):
+    def test_concrete_dropout(self):
```
Comment (wenkesj) [inline pyoneer/layers/dropout_test.py]: Can we test the regularization loss, as well? I.e. for different drop out rates, the regularization should change?

File: pyoneer/layers/dropout_impl.py (hunk 1 above)
Comment (wenkesj) [reply pyoneer/layers/dropout_impl.py]: For example, I see a case where someone supplies a constant tensor to the initial_rate argument and this blows up. Or maybe we should assert something in the constructor OR put it in the docs and have a test for it, if we want to keep this.

File: pyoneer/layers/dropout_impl.py (hunk 3 above)
Comment (wenkesj) [reply pyoneer/layers/dropout_impl.py]: Maybe it boils down to having a `ConcreteDropoutRegularizer` and a `ConcreteKernelRegularizer` or something. Ultimately the losses get summed together.

File: pyoneer/layers/dropout_impl.py (hunk 2 above)
Comment (wenkesj) [reply pyoneer/layers/dropout_impl.py]: If we decide to use the expansion operators, we should ensure that the versions of python that we support are explicit (this only works for 3.x not 2.x). I think the inspiration here is from the [keras/tf approach](https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/python/keras/layers/core.py#L168-L175)

File: pyoneer/layers/dropout_impl.py (hunk 1 above)
Comment (jimfleming) [reply pyoneer/layers/dropout_impl.py]: Definitely can use tf. I should probably add an op for the sigmoid inverse too. That’s generally useful.

File: pyoneer/layers/dropout_impl.py (hunk 3 above)
Comment (jimfleming) [reply pyoneer/layers/dropout_impl.py]: Yeah, I’ll look into that. The challenge is both regularizers need access to the current dropout rate and the downstream kernel. Any thoughts on API?

```
droput_layer = ConcreteDropout()
droput_regularizer = ConcreteDropoutRegularizer(dropout_layer.rate_logit)  # NOTE: rate_logit doesn’t exist until `build` has been called
kernel_regularizer = ConcretKernelRegularizer(dropout_layer.rate_logit)
```

File: pyoneer/layers/dropout_impl.py (hunk 5 above)
Comment (jimfleming) [reply pyoneer/layers/dropout_impl.py]: Hmm, I’m not opposed but I can’t think of a use case. When would you want only part of the layer?

File: pyoneer/layers/dropout_impl.py (hunk 2 above)
Comment (jimfleming) [reply pyoneer/layers/dropout_impl.py]: Yeah, this is lifted straight out of the Keras `Dropout` layer implementation. I like the expansion but we’d have to drop Python 2.7 support.

File: pyoneer/layers/dropout_test.py (hunk 6 above)
Comment (jimfleming) [reply pyoneer/layers/dropout_test.py]: Yep, this was rushed. It definitely needs more tests.

File: pyoneer/layers/dropout_impl.py (hunk 4 above)
Comment (jimfleming) [reply pyoneer/layers/dropout_impl.py]: I agree. I can make it an argument. This is how Yarin implemented it but I can keep 0.1 as the default.

File: pyoneer/layers/dropout_impl.py (hunk 3 above)
Comment (wenkesj) [reply pyoneer/layers/dropout_impl.py]: Hmm, maybe this is something that should be added later. I'm not sure about the implementation details from the paper, but that may help me understand it better.

File: pyoneer/layers/dropout_impl.py (hunk 5 above)
Comment (wenkesj) [reply pyoneer/layers/dropout_impl.py]: That is a good point! I can't really see passed it's only use case here.

**Comment by djsaunde** [comment]
Can we get this merged in?
